model_analyzer/result/concurrency_search.py

```diff
@@ -111,31 +113,56 @@ def _has_objective_gain_saturated(self) -> bool:
     def _calculate_gain(self) -> float:
         first_rcm = self._run_config_measurements[
             -THROUGHPUT_MINIMUM_CONSECUTIVE_CONCURRENCY_TRIES]
-        best_rcm = max(self._run_config_measurements[
-            -THROUGHPUT_MINIMUM_CONSECUTIVE_CONCURRENCY_TRIES:])
 
-        gain = first_rcm.compare_measurements(best_rcm)
+        best_rcm = self._get_best_rcm()
+
+        # These cover the cases where we don't get a result from PA
+        if not first_rcm and not best_rcm:
+            return 0
+        if not first_rcm:
+            return 1
+        elif not best_rcm:
+            return -1
+        else:
+            gain = first_rcm.compare_measurements(best_rcm)
 
         return gain
 
+    def _get_best_rcm(self) -> Optional[RunConfigMeasurement]:
+        # Need to remove entries (None) with no result from PA before sorting
+        pruned_rcms = [
+            rcm for rcm in self._run_config_measurements[
+                -THROUGHPUT_MINIMUM_CONSECUTIVE_CONCURRENCY_TRIES:] if rcm
+        ]
+        best_rcm = max(pruned_rcms) if pruned_rcms else None
+
+        return best_rcm
+
     def _was_constraint_violated(self) -> bool:
         for i in range(len(self._run_config_measurements) - 1, 1, -1):
             if self._at_constraint_failure_boundary(i):
                 self._last_failing_concurrency = self._concurrencies[i]
                 self._last_passing_concurrency = self._concurrencies[i - 1]
                 return True
 
-        if not self._run_config_measurements[0].is_passing_constraints():
+        if self._run_config_measurements[
+                0] and not self._run_config_measurements[
+                    0].is_passing_constraints():
             self._last_failing_concurrency = self._concurrencies[i]
             self._last_passing_concurrency = 0
             return True
         else:
             return False
 
     def _at_constraint_failure_boundary(self, index: int) -> bool:
-        at_failure_boundary = not self._run_config_measurements[
+        if not self._run_config_measurements[
+                index] or not self._run_config_measurements[index - 1]:
+            return False
+
+        at_failure_boundary = not self._run_config_measurements[  # type: ignore
```

Comment:
are 2 type ignores needed? Makes this hard to read

Reply:
Grrr...yes, it's really annoying, but I can't figure out any other way to do this.

Reply:
optional measurements have always caused us problems. These type hints point to part of our design that we should try to clean up

Reply:
Well, tell PA to stop returning no measurement and this wouldn't need to be optional!